src/condor_starter.V6.1/starter.cpp

```diff
@@ -2077,10 +2081,309 @@ Starter::jobEnvironmentCannotReady(int status, const struct UnreadyReason & urea
+			// contents.  Instead, let's just redirect the streams to a
+			// temporary file.
+			//
+			int dev_null_fd = open("/dev/null", O_RDONLY);
+			if( dev_null_fd == -1) {
+				diagnosticResultAd.InsertAttr( "Result", "Error - Unable to open /dev/null" );
+				SEND_REPLY_AND_CONTINUE_CONVERSATION;
+			}
+			char tmpl[] = "XXXXXX";
+			int log_file_fd = condor_mkstemp(tmpl);
+			if( log_file_fd == -1 ) {
+				diagnosticResultAd.InsertAttr( "Result", "Error - Unable to open temporary file" );
+				SEND_REPLY_AND_CONTINUE_CONVERSATION;
+			}
+
+			int the_redirects[3] = {dev_null_fd, log_file_fd, log_file_fd};
+
+			condor::dc::AwaitableDeadlineReaper logansRun;
+
+			std::vector< std::string > diagnostic_args;
+			diagnostic_args.push_back( diagnostic_path.string() );
+
+			Env diagnostic_env;
+			// This should include CONDOR_CONFIG.
+			diagnostic_env.Import();
+
+			std::string condor_bin;
+			param( condor_bin, "BIN" );
+			if(! condor_bin.empty()) {
+				diagnostic_env.SetEnv("_CONDOR_BIN", condor_bin);
+			}
+
+			// Windows doesn't have getppid().
+			std::string _condor_starter_pid;
+			formatstr( _condor_starter_pid, "%d", getpid() );
+			diagnostic_env.SetEnv("_CONDOR_STARTER_PID", _condor_starter_pid);
+
+			OptionalCreateProcessArgs diagnostic_process_opts;
+			int spawned_pid = daemonCore->CreateProcessNew(
+				diagnostic_path.string(),
+				diagnostic_args,
+				diagnostic_process_opts
+					.reaperID(logansRun.reaper_id())
+					.priv(PRIV_USER_FINAL)
+					.std(the_redirects)
+					.env(& diagnostic_env)
+			);
+
+			logansRun.born( spawned_pid, 20 ); // ... seconds of careful research
+
+			while( logansRun.living() ) {
+				auto [pid, timed_out, status] = co_await logansRun;
+				if( timed_out ) {
+					diagnosticResultAd.InsertAttr( "Result", "Error - Timed Out" );
+				} else {
+					std::string log_bytes;
+
+					size_t bytes_read = 0;
+					size_t total_bytes_read = 0;
+
+					const size_t BUFFER_SIZE = 32768;
+					char buffer[BUFFER_SIZE];
+					if( lseek(log_file_fd, 0, SEEK_SET) == (off_t)-1 ) {
+						diagnosticResultAd.InsertAttr( "Result", "Error - failed to lseek() log" );
+						diagnosticResultAd.InsertAttr( "ExitStatus", status );
+						SEND_REPLY_AND_CONTINUE_CONVERSATION;
+					}
+
+					while( (bytes_read = full_read(log_file_fd, (void *)buffer, BUFFER_SIZE)) != 0 ) {
+						log_bytes.insert(total_bytes_read, buffer, bytes_read);
+						total_bytes_read += bytes_read;
+						if( bytes_read != BUFFER_SIZE ) {
+							break;
+						}
+					}
+
+
+					char * base64Encoded = condor_base64_encode(
+						(const unsigned char *)log_bytes.c_str(),
+						total_bytes_read, false
+					);
+					diagnosticResultAd.InsertAttr( "Contents", base64Encoded );
+					free( base64Encoded );
+
+
+					diagnosticResultAd.InsertAttr( "Result", "Completed" );
+					diagnosticResultAd.InsertAttr( "ExitStatus", status );
+				}
+			}
+		}
+	}
+
+
+	//
+	// Reply.
+	//
+	SEND_REPLY_AND_CONTINUE_CONVERSATION;
+}
```

Comment:
@JaimeFrey The `CreateProcessNew()` call in the middle of this function _should_ be safe (and do nothing as root), but given the recent brouhaha, I figure a second careful check wouldn't hurt.

Reply:
It doesn't look like you're in PRIV_ROOT for any of this code, and the child process is PRIV_USER_FINAL. There doesn't seem to be much to say w.r.t. the whole doing-stuff-as-root thing.

Reply:
That's what I thought, but I'm glad to get it confirmed.